File: src/api/ai_generation.ts
import { invoke } from '@tauri-apps/api/core';
import type { GenerateRequest, GenerateResponse, AiGenerationHistory } from '../types/ai_generation';

export async function aiGenerate(request: GenerateRequest): Promise<GenerateResponse> {
  return invoke('ai_generate', { request });
}

/// 流式生成：不返回内容（通过事件推送），仅返回调用是否成功
export async function aiGenerateStream(request: GenerateRequest): Promise<void> {
  return invoke('ai_generate_stream', { request });
}

export async function listAiGenerationHistory(chapterId: string): Promise<AiGenerationHistory[]> {
  return invoke('list_ai_generation_history', { chapterId });
}

export async function getAiGenerationHistory(id: string): Promise<AiGenerationHistory> {
  return invoke('get_ai_generation_history', { id });
}

export async function deleteAiGenerationHistory(id: string): Promise<void> {
  return invoke('delete_ai_generation_history', { id });
}

export async function deleteAiGenerationHistoryByChapter(chapterId: string): Promise<number> {
  return invoke('delete_ai_generation_history_by_chapter', { chapterId });
}

File: src-tauri/src/commands/ai_generation.rs
use serde::{Deserialize, Serialize};
use sqlx::SqlitePool;
use tauri::{AppHandle, Manager, State};

use crate::error::AppError;
use crate::models::ai_generation::{AiGenerationHistory, GenerationParameters};
use crate::services::ai_generation_service;

#[derive(Deserialize)]
pub struct GenerateRequest {
    pub chapter_id: String,
    pub mode: String,
    pub user_instruction: Option<String>,
    pub parameters: Option<GenerationParameters>,
}

#[derive(Serialize)]
pub struct GenerateResponse {
    pub content: String,
    pub history: AiGenerationHistory,
}

#[tauri::command]
pub async fn ai_generate(
    app: AppHandle,
    pool: State<'_, SqlitePool>,
    request: GenerateRequest,
) -> Result<GenerateResponse, AppError> {
    let app_data_dir = app.path().app_data_dir().map_err(|e| {
        AppError::Internal(format!("Failed to get app data dir: {}", e))
    })?;

    let (content, history) = ai_generation_service::generate(
        pool.inner(),
        &app_data_dir,
        &request.chapter_id,
        &request.mode,
        request.user_instruction.as_deref(),
        request.parameters,
    ).await?;

    Ok(GenerateResponse { content, history })
}

/// 流式 AI 生成命令
///
/// 不返回生成内容（通过事件 `ai-generation-chunk` / `ai-generation-done` 推送），
/// 仅返回调用是否成功启动。生成内容、历史记录均通过事件传递。
#[tauri::command]
pub async fn ai_generate_stream(
    app: AppHandle,
    pool: State<'_, SqlitePool>,
    request: GenerateRequest,
) -> Result<(), AppError> {
    let app_data_dir = app.path().app_data_dir().map_err(|e| {
        AppError::Internal(format!("Failed to get app data dir: {}", e))
    })?;

    ai_generation_service::generate_stream(
        pool.inner(),
        &app,
        &app_data_dir,
        &request.chapter_id,
        &request.mode,
        request.user_instruction.as_deref(),
        request.parameters,
    )
    .await
}

#[tauri::command]
pub async fn list_ai_generation_history(
    pool: State<'_, SqlitePool>,
    chapter_id: String,
) -> Result<Vec<AiGenerationHistory>, AppError> {
    ai_generation_service::list_history(pool.inner(), &chapter_id).await
}

#[tauri::command]
pub async fn get_ai_generation_history(
    pool: State<'_, SqlitePool>,
    id: String,
) -> Result<AiGenerationHistory, AppError> {
    ai_generation_service::get_history(pool.inner(), &id).await
}

#[tauri::command]
pub async fn delete_ai_generation_history(
    pool: State<'_, SqlitePool>,
    id: String,
) -> Result<(), AppError> {
    ai_generation_service::delete_history(pool.inner(), &id).await
}

#[tauri::command]
pub async fn delete_ai_generation_history_by_chapter(
    pool: State<'_, SqlitePool>,
    chapter_id: String,
) -> Result<u64, AppError> {
    ai_generation_service::delete_history_by_chapter(pool.inner(), &chapter_id).await
}
